File: Stack/Leetcode/Medium/RemoveKDigits/RemoveKDigits.java
// https://leetcode.com/problems/remove-k-digits/

import java.util.ArrayDeque;
import java.util.Deque;


public class RemoveKDigits {

    public String removeKdigits(String num, int k) {
        Deque<Character> st = new ArrayDeque<>();
        for (int i = 0; i < num.length(); i++) {
            char ch = num.charAt(i);
            while (!st.isEmpty() && k > 0 && st.peek() > ch) {
                st.pop();
                k--;
            }
            if (!st.isEmpty() || ch != '0') {
                st.push(ch);
            }
        }
        while (!st.isEmpty() && k > 0) {
            st.pop();
            k--;
        }
        StringBuilder ans = new StringBuilder();
        for (char c : st) {
            ans.append(c);
        }

        return ans.length() == 0 ? "0" : ans.reverse().toString();
    }
}

File: Stack/Leetcode/Medium/RemoveKDigits/RemoveKDigits.cpp
// https://leetcode.com/problems/remove-k-digits/
#include<bits/stdc++.h>
using namespace std;
string removeKdigits(string num, int k) {
        stack<char> st;
        for(int i=0; i<num.length(); i++) {
            while(!st.empty() && k > 0 && st.top() > num[i]) {
                st.pop();
                k--;
            }
            if(!st.empty() ||  num[i] != '0') st.push(num[i]);
        }
        while(!st.empty() && k > 0) {
            st.pop();
            k--;
        }
        string ans ;
        while(!st.empty()) {
            ans.push_back(st.top());
            st.pop();
        }

        reverse(ans.begin(), ans.end());

        return ans.length() == 0 ? "0" : ans;
    }